man/dbdscr.1
.TH DBDSCR 1 "12 September 1995"
.FB
.SH NAME
dbdscr \- define a \fIFirstBase\fP database screen dictionary
.SH SYNOPSIS
.B dbdscr
[\fB-d\fP \fIdatabase\fP] [\fB-s\fP \fIscreen\fP]
.SH DESCRIPTION
.I Dbdscr
allows definition and modification of a 
.I FirstBase
database screen dictionary. 
The created dictionary can later be used by
\fIdbedit\fP(1)
to effectively mask off and/or reorder the fields of a database.
In other words, a screen dictionary can be used to provide a logical
to physical map for the database fields.
.PP
A database screen dictionary consists of 
a series of screen descriptors. Each screen descriptor 
contains a
.I field_name,
and an optional
.I default-string.
.PP
Like all other \fIFirstBase\fP dictionary editors, \fIdbdscr\fP(1) has a set of commands
that allow screen oriented modification of the screen dictionary.
The following is a summary of the commands accepted by 
\fIdbdscr\fP(1):
.TP 7
.B <CTL>-H
displays a summary of these commands.
.TP 7
.B ?[dfhiN]
displays help on the database, formulas,
user defined help files, index, or field N.
.TP 7
.B <RETURN>
(return, or enter) pages forward one page in the database
dictionary (same as 'f').
Wrapping occurs at end of file.
.TP 7
.B @
go into auto add mode.
.TP 7
.B #
allow changes to line number #.
.TP 7
.B -
standard END key \(em exits program and keeps all changes.
.TP 7
.B <CTL>-X
aborts the program. The effect is to exit without
overwriting the file, meaning all changes are lost.
.TP 7
.B b
pages backward one page in the database screen dictionary. 
Wrapping occurs at beginning of file.
.TP 7
.B d
delete function. Allows deletion of a single dictionary entry.
All entries below deleted entry slide up one slot.
.TP 7
.B i
insert function. Inserts a standard dictionary entry before any given
field.
All entries below inserted entry slide down one slot.
.PP
A list of simple editor commands appears on the 24th line of the 
screen at all times, with help available by using the standard help 
signal (<CTL>-H).
Complete interactive error checking is done on the contents of
all screen descriptors.
.PP
Upon successful completion of a screen dictionary, the user is given a
chance to run the database editor. All arguments given
to \fIdbdscr\fP(1) are passed on to \fIdbedit\fP(1).
.SH FILES
.PD 0
.TP 10
dbase
default name for the database screen
.TP 10
*.sdict
dictionary for database screen.
.PD
.SH SEE ALSO
dbedit(1), dbdbas(1)
.PP
.I FirstBase User's Guide and Reference Manual
.br
.SH DIAGNOSTICS
\fIDbdscr\fP(1)
will allow editing and modifying of any screen dictionary at any time.
These dictionary files should be write protected for complete protection.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbedit/dbdscr.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbdscr.c,v 9.1 2001/02/16 19:41:42 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char sccsid[] = "@(#)dbdscr.c	8.3 05 Apr 1996 FB";
#endif

/* 
 *  dbdscr.c - screen editor for screen dictionary (sdict) file
 */

#include <fb.h>
#include <fb_vars.h>
#include <dbd.h>

static int dot = 0;			/* current fb_field */
static int lastitem = 0;		/* last list item */

static fb_upd **itemlist = NULL;	/* for list of items */

static int col[8] = {1, 6, 18, 22, 30};	/* edit columns (screen) */

static fb_database *hp;
static char sdict[FB_MAXNAME];		/* for sdict filename */

static char *HLP_DBDSCR = "dbdscr.hlp";

#define SHOWP 2				/* ending of shown screen columns */

#if FB_PROTOTYPES
static edit_sdict();
static fillscreen();
static escreen();
static void fdelete();
static void finsert();
static fadder();
static eitem();
static fput();
static new_sdict();
static headers();
static usage();
static read_sdict(char *fname, fb_upd **itemlist);
#else /* FB_PROTOTYPES */
static edit_sdict();
static fillscreen();
static escreen();
static void fdelete();
static void finsert();
static fadder();
static eitem();
static fput();
static new_sdict();
static headers();
static usage();
static read_sdict();
#endif /* FB_PROTOTYPES */

extern short int cdb_locklevel;
extern short int cdb_askgen;

extern char *cdb_EDIT;

/* 
 *  dbdscr - main driver
 */
 
   main(argc, argv)
      int argc;
      char *argv[];
      
      {
         int st, i, p;

	 cdb_locklevel = -1;
         fb_getargs(argc, argv, FB_ALLOCDB);
	 hp = cdb_db;
	 fb_initdbd(hp);
         if (fb_getd_dict(hp) == FB_ERROR)
            fb_xerror(FB_BAD_DICT, hp->ddict, NIL);            
         else
            for (i = 0; i < hp->nfields; i++)
               fb_nounders(cdb_keymap[i]);
	 fb_scrhdr(hp, "Define Screen");
         headers();
	 fb_rootname(sdict, hp->dbase);
	 if ((p = fb_testargs(argc, argv, "-s")) > 0){
	    if (++p >= argc)
	       usage();
	    if (!equal(argv[p], "-"))	/* get around the -s - menu call */
	       fb_rootname(sdict, argv[p]);
	    }
	 strcat(sdict, ".sdict");
         st = edit_sdict(sdict);
         if (st == FB_END){
            new_sdict(sdict);
	    if (cdb_askgen)
	       fb_chainto("Run Dbedit", cdb_EDIT, argv);
	    }
         fb_ender();
      }

/* 
 *  edit_sdict - edit an dbase.sdict or allow creation of one 
 */
 
   static edit_sdict(p)
      char *p;
      
      {
         int st, i;

	 itemlist =
	     (fb_upd **) fb_malloc((hp->nfields+1) * sizeof(fb_upd *));
	 			/* allocate enough ptrs for all fields */
	 if ((lastitem = read_sdict(p, itemlist)) == FB_ERROR)
	    lastitem = 0;
	 for (i = lastitem + 1; i < hp->nfields ; i++)
	    itemlist[i] = NULL;
         dot = 1;
         fillscreen();
         while ((st = escreen()) != FB_END)
            if (st == FB_ABORT)
               break;
            else if (st >= dot && st <= dot + 9 && st <= lastitem){
               fb_cx_push_env("6D", CX_KEY_SELECT, NIL);
               eitem(st, 0);
               fb_cx_pop_env();
               }
            else if (st >= 1 && st <= lastitem){
               dot = fb_onpage(st);
               fillscreen();
               }
            else
               fb_screrr("Invalid Input");
         return(st);
      }

/* 
 *  fillscreen - fill a screen with items
 */
 
   static fillscreen()
      {
         int j, i, p;
         
         fb_move(4, 1); fb_clrtobot();
         j = dot + 9;
         for (i = dot; i <= j && i <= lastitem; i++)
            for (p = 0; p <= SHOWP; ++p)
               fput(i, p);
	 fb_infoline();
         fb_cx_set_viewpage(dot);
      }
            
/* 
 *  escreen - set up a screen, accept commands for screen control 
 */
 
   static escreen()
      {
         int st;
	 char com[6];
         
         for (;;){
            fb_cx_push_env("E1", CX_DBD_SELECT, NIL);
            if (lastitem > 10)
               fb_cx_add_buttons("Ge");
            fb_cx_add_buttons("XH");
            st = fb_anychange(com);
            fb_cx_pop_env();
            switch (st) {
               case FB_ABORT:
	          if (fb_mustbe('y', "Really Quit? (y/n) ",
                        cdb_t_lines, 1) == FB_AOK){
                     fb_screrr("Screen Dictionary File Unchanged");
		     return(st);
		     }
		  break;
               case ADDMODE:
	          if (lastitem < hp->nfields)
                     fadder();
		  else
		     continue;
                  break;
               case DELMODE:
	          if (lastitem > 0)
                     fdelete();
		  else
		     continue;
                  break;
               case FB_END:
                  return(st);
	       case HELPMODE:
	          fb_move(2, 1); fb_clrtobot();
		  fb_help(com, hp);
		  fb_move(4,1); fb_clrtobot();
		  headers();
		  break;
	       case INSERTMODE:
	          if (lastitem > 0 && lastitem < hp->nfields)
	             finsert();
		  else
		     continue;
		  break;
               case PAGEF:
                  dot = dot + 10;
                  if (dot > lastitem)
                     dot = 1;
                  break;
               case PAGEB:
                  dot = dot - 10;
                  if (dot < 1)
                     dot = fb_onpage(lastitem);
                  break;
	       case FB_QHELP: 
	          fb_move(2, 1); fb_clrtobot();
	          fb_fhelp(HLP_DBDSCR);
		  headers();
		  break;
	       case FB_AOK:
                  return(atoi(com));
               default:
                  return(st);
               }
            fillscreen();
            }
      }

/* 
 *  fdelete - delete an update item 
 */
 
   static void fdelete()
      {
         int st, n;
         
         fb_move(cdb_t_lines, 1);
         fb_clrtoeol();
         fb_printw("Delete what number ? (#,-) ");
         st = fb_input(cdb_t_lines, 28, 3, 0, FB_INTEGER, (char *) &n,
            FB_ECHO,FB_OKEND,FB_CONFIRM);
         if (st == FB_END || st == FB_ABORT)
            return;
	 else if (n < 1 || n > lastitem)
	    return;
	 if (itemlist[n] != NULL)
	    fb_free((char *) itemlist[n]);
         for (; n < lastitem; n++)
            itemlist[n] = itemlist[n + 1];
         lastitem--;
      }

/* 
 *  finsert - insert a defaulted item before a given item 
 */
 
   static void finsert()            
      {
         int st, n, rn;

	 fb_fmessage("Insert before what number (#,-) ? ");
         st = fb_input(cdb_t_lines, 36, 3, 0, FB_INTEGER, (char *) &n, FB_ECHO,
            FB_OKEND, FB_CONFIRM);
         if (st == FB_END || st == FB_ABORT)
            return;
	 else if (n < 1 || n > lastitem)
	    return;
	 rn = n;
         for (n = lastitem; n >= rn; n--)		/* go backwards */
            itemlist[n+1] = itemlist[n];
	 itemlist[rn] = (fb_upd *) fb_malloc(sizeof(fb_upd));
	 itemlist[rn]->fp = NULL;		/* default to first fb_field */
	 itemlist[rn]->fv = NULL;
         lastitem++;
      }

/* 
 *  fadder - go into auto add mode 
 */
 
   static fadder()
      {
         int st;

         dot = fb_onpage(lastitem);
	 if ((lastitem + 1) < (dot + 10))
            fillscreen();
         for (;lastitem < hp->nfields;){
	    itemlist[++lastitem] = (fb_upd *) fb_malloc(sizeof(fb_upd));
	    itemlist[lastitem]->fp = NULL;
	    itemlist[lastitem]->fv = NULL;
            if (lastitem >= dot + 10){
               dot += 10;
               fb_move(4, 1); fb_clrtobot(); fb_infoline();
               }
            fb_cx_push_env("6D", CX_KEY_SELECT, NIL);
            st = eitem(lastitem, 1);
            fb_cx_pop_env();
            if (st == FB_END){
	       fb_free((char *) itemlist[lastitem--]);
               break;
	       }
            }
      }

/* 
 *  eitem - edit an entire item 
 */
 
   static eitem(it, new)
      int it, new;

      {
         int row, i, st;
         char *fb_trim(), temp[FB_MAXLINE];
	 fb_field *fp = NULL;

         fb_move(cdb_t_lines, 1); fb_clrtoeol();
         if ((row = (it % 10)) == 0)
            row = 10;
         row = row * 2 + 2;
	 if (itemlist[it] == NULL)
	    fb_xerror(FB_BAD_DICT, hp->idict, "bad itemlist");
         if (new == 1)
            fput(it, 0);

         /* 
	  *  get id 
	  */
	  
         for (;;){
	    if (new)
               fb_fmessage("Enter FieldName, -=End");
	    else
               fb_fmessage("Enter FieldName, <CTL>-X=Skip");
            st = fb_input(row, col[1], FB_TITLESIZE, 0, FB_ALPHA, 
                 temp, FB_ECHO, FB_OKEND, FB_CONFIRM);
            if (st == FB_END && new == 1){
               fb_move(row, 1); 
               fb_clrtoeol();
               return(FB_END);
               }
            else if (st == FB_END)
               st = FB_DEFAULT;
            if (st == FB_DEFAULT || (st == FB_ABORT && new == 1)){
	       if ((it - 1) < hp->nfields)
	          fp = cdb_kp[it - 1];
	       else
	          fp = cdb_kp[0];
	       }
            else if (st == FB_ABORT)
               break;
            else if (st == FB_AOK){
               fb_trim(temp);
	       fp = fb_findfield(temp, hp);
	       }
	    if (fp == NULL){
	       fb_screrr("Cannot locate that field. Sorry.");
	       continue;
	       }
	    else if (it == 1 && (fp->type == FB_FORMULA || fp->type==FB_LINK)){
	       fb_screrr("Must use a non-formula/link field as first field!");
	       continue;
	       }
	    for(i = 1; i <= lastitem; i++)
	       if (itemlist[i]->fp == fp)
	          break;
	    if (i <= lastitem){
	       fb_screrr("That field is already in this screen dictionary!");
	       continue;
	       }
            itemlist[it]->fp = fp;
	    break;
            }
         fput(it, 1);
	 
	 /* 
	  *  get override default 
	  */
	  
         for(;;){
	    if (itemlist[it]->fp->type == FB_FORMULA ||
	        itemlist[it]->fp->type == FB_LINK)
	       break;
	    if (new)
	       fb_fmessage("Enter Default Value");
	    else
	       fb_fmessage("Enter Default Value, <CTL>-X=Skip");
            st = fb_input(row, col[2], 15, 0, FB_ALPHA,temp, FB_ECHO,
               FB_NOEND, FB_CONFIRM);
            if (st == FB_ABORT && new != 1)
               strcpy(temp, itemlist[it]->fv);
	    else if (st == FB_DEFAULT || st == FB_ABORT)
	       strcpy(temp, " ");
	    else if (st == FB_AOK){
	       fb_trim(temp);
	       if (temp[0] == NULL)
	          strcpy(temp, "_");
	       fb_underscore(temp, 1);
	       }
	    fb_mkstr(&(itemlist[it]->fv), temp);
            break;
            }
         fput(it, 2);
         return(FB_AOK);
      }

/* 
 *  fput - fb_put an entire item to a constant screen location 
 */
 
   static fput(i, pos)
      int i, pos;
      
      {
         int row;
         
         if ((row = (i % 10)) == 0)
            row = 10;
         row = row * 2 + 2;
         fb_move(row, col[pos]);
         switch (pos){
            case 0: fb_printw("%3d)", i); break;
            case 1: 
	       if (itemlist[i]->fp != NULL)
	          fb_printw("%-10s", itemlist[i]->fp->id);
	       else
	          fb_printw("<empty>");
	       break;
	    case 2:
	       fb_clrtoeol();
	       if (itemlist[i]->fv != NULL)
	          fb_printw(FB_FSTRING, itemlist[i]->fv); 
	       break;
            }
      }

/* 
 *  new_sdict - printout a new dictionary list of itemlist 
 */
 
   static new_sdict(p)
      char *p;
      
      {
         int i;
	 FILE *fs, *fb_mustfopen();
	 
	 fs = fb_mustfopen(p, "w");
	 for (i = 1; i <= lastitem; i++){
	    if (itemlist[i]->fp != NULL)
	       fprintf(fs, "%s", fb_underscore(itemlist[i]->fp->id, 1));
	    if (itemlist[i]->fv != NULL && 
	          strlen(fb_trim(itemlist[i]->fv)) > 0)
	       fprintf(fs, " -d %s", fb_underscore(itemlist[i]->fv, 1));
	    fprintf(fs, "\n");
	    }
	 fclose(fs);
      }
      
/* 
 *  headers - print header at top of screen 
 */
 
   static headers()
      {
         fb_move(2, 1); fb_clrtoeol();
	 fb_move(3, 1); fb_clrtoeol();
         fb_move(2,col[1]);   fb_printw("Field");
         fb_move(3,col[1]);   fb_printw("=====");
	 fb_scrstat2(sdict);
	 fb_move(2,col[2]); fb_printw("Default");
	 fb_move(3,col[2]); fb_printw("=======");
/*	 fb_move(2,col[3]+2); fb_printw("Size");
*	 fb_move(3,col[3]+2); fb_printw("====");
*	 fb_move(2,col[4]);   fb_printw("Update String");
*	 fb_move(3,col[4]);   fb_printw("=============");
*/	 
      }
      
/* 
 *  usage message 
 */
 
   static usage()
      {
         fb_xerror(FB_MESSAGE, "usage: dbdscr [-d dbase] [-s index]", NIL);
      }

/* 
 *  read_sdict - read in an existing sdict file for editing 
 */
 
   static read_sdict(fname, itemlist)
      char *fname;
      fb_upd **itemlist;
      
      {
         char line[FB_MAXLINE], word[FB_MAXLINE];
	 fb_upd *up;
	 fb_field *fp;
	 int p, q;
	 FILE *fs;
	 
	 if ((fs = fopen(fname, "r")) == NULL)
	    return(FB_ERROR);
	 q = 0;
	 while (fgets(line, FB_MAXLINE, fs) != NULL){
	    if (line[0] == '%')
	       break;
	    q++;
	    p = fb_getword(line, 1, word);
	    fb_underscore(word, 0);
	    if ((fp = fb_findfield(word, cdb_db)) == NULL)
	       fb_xerror(FB_BAD_DICT, word, NIL);
	    up = (fb_upd *) fb_malloc(sizeof(fb_upd));
	    up->fp = fp;
	    up->fv = NULL;
	    itemlist[q] = up;
	    
	    p = fb_getword(line, p, word);
	    if (equal(word, "-d")){
	       p = fb_getword(line, p, word);
	       fb_mkstr(&(up->fv), word);
	       }
	    else
	       fb_mkstr(&(up->fv), " ");
	    }
	 fclose(fs);
	 return(q);
      }
